File: constructi-backend/src/main/java/com/constructi/service/SubtaskService.java
package com.constructi.service;

import com.constructi.DTO.SubtaskRequestDTO;
import com.constructi.DTO.SubtaskResponseDTO;

import java.util.List;

public interface SubtaskService {
    SubtaskResponseDTO createSubtask(SubtaskRequestDTO subtaskRequestDTO);

    SubtaskResponseDTO updateSubtask(Long subtaskId, SubtaskRequestDTO subtaskRequestDTO);

    void deleteSubtask(Long subtaskId);

    List<SubtaskResponseDTO> getAllSubtasks();

    SubtaskResponseDTO getSubtaskById(Long subtaskId);

    List<SubtaskResponseDTO> getSubtasksByParentTaskId(Long parentTaskId);

    SubtaskResponseDTO approveSubtask(Long subtaskId);

}

File: constructi-frontend/src/app/modules/subtask/subtask.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class SubtaskService {
  private apiUrl = 'http://localhost:8086/subtasks';

  constructor(private http: HttpClient) {}

  createSubtask(subtask: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/add`, subtask);
  }

  getAllSubtasks(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}`);
  }

  getSubtaskById(subtaskId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${subtaskId}`);
  }

  updateSubtask(subtaskId: number, subtask: any): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/update/${subtaskId}`, subtask);
  }

  prolongSubtask(taskId: number, newEndDate: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/prolong/${taskId}`, null, {
      params: { newEndDate }
    });
  }

  deleteSubtask(subtaskId: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/delete/${subtaskId}`, { responseType: 'text' });
  }

  getSubtasksByParentTaskId(parentTaskId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/parent/${parentTaskId}`);
  }

  approveSubtask(subtaskId: number): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/approve/${subtaskId}`, {});
  }
}
